### ISSUE TITLE
Add matplotlib support to PyCharm

### ISSUE THREAD

**Issue by ghost** [donkirkby/live-py-plugin#52]
Would you be interested in adding support for [matplotlib](http://matplotlib.org/)?

Currently, examples using that library prints nothing.

It would be neat if the `matplotlib.figure.Figure` called `p` below were shown in a buffer.

``` python
import seaborn as sns
sns.set(style="ticks")

# Load the example dataset for Anscombe's quartet
df = sns.load_dataset("anscombe")

# Show the results of a linear regression within each dataset
g = sns.lmplot(x="x", y="y", col="dataset", hue="dataset", data=df,
               col_wrap=2, ci=None, palette="muted", size=4,
               scatter_kws={"s": 50, "alpha": 1})#.getfig()

p = g.fig
```

Example from here: [Anscombe’s quartet](http://stanford.edu/~mwaskom/software/seaborn-dev/examples/anscombes_quartet.html)

**Comment by endrebak** [comment]
The simplest thing that would work is probably to test whether object is a `matplotlib.figure.Figure` object, save as `/tmp/live-py/graph_N.png` and make that picture available. Emacs has a picture-mode, I'm sure Eclipse has one too.

**Comment by donkirkby** [comment]
Thanks for the suggestion, I could probably do something similar to the way I show turtle graphics.

For now, here's the technique I use. I make the live coding block write to a test file, and then open that file in an image viewer. My image viewer automatically updates its display each time the code runs, and I can even keep the window on top of my Eclipse editor.

Here's a trivial example:

    from matplotlib import pyplot as plt
    
    
    def draw(filename):
        x = [1, 4, 5, 8]
        y = [2, 4, 7, 3]
    
        plt.plot(x, y)
        plt.savefig(filename)  # write image to file
        plt.clf()
    
    if __name__ == '__live_coding__':
        draw('scratch.png')

I don't think you meant picture-mode in Emacs, that's just a different way to move around a text buffer. It sounds like Emacs does [support images][images], though. You might even be able to do something similar to the code above, and display the image in Emacs.

[images]: http://www.gnu.org/software/emacs/manual/html_node/elisp/Images.html

**Comment by donkirkby** [comment]
I could build in support for matplotlib by replacing [plt.show()] with a call to [plt.savefig()] and a `BytesIO` object. Then the bytes from the `BytesIO` object could get sent back to the IDE for display.

It might also be interesting to build a set of tutorials that display a goal plot and the plot from your code. The objective is to fix the code to produce the target plot. Might be able to make a learning tool inspired by [learngitbranching.js.org]. Call it matplotlearn?

> fname:: A string containing a path to a filename, or a Python file-like object

[plt.savefig()]: http://matplotlib.org/api/pyplot_api.html#matplotlib.pyplot.savefig
[plt.show()]: http://matplotlib.org/api/pyplot_api.html#matplotlib.pyplot.show
[learngitbranching.js.org]: http://learngitbranching.js.org/

**Comment by brandm** [comment]
"I could build in support for matplotlib [...]"

That would be awesome. Right now I could use it well for an ordinary drawing.

How is the plan to show the image in Emacs? My suggestion is to keep the current output window side by side with the Python source in one frame and to open a different frame exclusively for the image, also for convenient frame border resize. I just did a small experiment for Emacs into this direction to transfer a PNG from a shell command through stdout without involving a regular file to show it in a new frame:

```elisp
(defvar live-py-image-buffer)
(let ((png-size8x8-color00ff00ff
       (concat "\\x89PNG\\x0d\\x0a\\x1a\\x0a\\x00\\x00\\x00\\x0d"
               "IHDR\\x00\\x00\\x00\\x08\\x00\\x00\\x00\\x08\\x08\\x06"
               "\\x00\\x00\\x00\\xc4\\x0f\\xbe\\x8b\\x00\\x00\\x00\\x12"
               "IDAT\\x78\\xda\\x63\\x64\\xf8\\x0f\\x84\\x78"
               "\\x00\\xe3\\xc8\\x50\\x00\\x00\\x15\\x1b\\x0f"
               "\\xf9\\xee\\x99\\xd6\\x69\\x00\\x00\\x00\\x00"
               "IEND\\xae\\x42\\x60\\x82")))
  (set (make-local-variable 'live-py-image-buffer)
       (concat "*live-py-image_"
               (file-name-nondirectory (buffer-file-name))
               "_"
               (make-temp-name "")
               "*"))
  (shell-command (concat "printf -- '" png-size8x8-color00ff00ff "'")
                 live-py-image-buffer)
  (with-current-buffer live-py-image-buffer
    (setq buffer-read-only 1)
    (image-mode)
    (make-frame)))
```

"inspired by [...] learngitbranching.js.org. Call it matplotlearn?"

learnmatplotlib would be better aligned with learnx and with the library name matplotlib.

**Comment by donkirkby** [comment]
Thanks for the info, @brandm, I didn't know about Emacs [image-mode]. That means we might also be able to support turtle graphics in Emacs for issue #80, at least in a GUI window.

As for how to display the image, I'm tempted to just display the source code and one of the buffers (program state, matplotlib, or turtle) side by side, then let the user create a [separate frame] if they want to. They could also just switch back and forth between the visible buffers, so we have to pay attention to which buffer is visible when we update the display.

[image-mode]: https://www.emacswiki.org/emacs/ImageMode
[separate frame]: https://www.gnu.org/software/emacs/manual/html_node/emacs/Creating-Frames.html

**Comment by brandm** [comment]
"[...] about Emacs image-mode."

For completeness: Emacs 25 built with GTK supports also Xwidgets which I guess could be used for number sliders like Bret Victor used them:
http://www.youtube.com/watch?v=PUv66718DII&t=4m17s

**Comment by donkirkby** [comment]
I did a little work on adding sliders to the [Eclipse] version, but I didn't make much progress. Maybe it will be easier in PyCharm or Emacs. Sliders are great for visual stuff like turtle graphics.

[Eclipse]: https://github.com/donkirkby/live-py-plugin/blob/master/plugin/src/live_py/ValueSlider.java

**Comment by brandm** [comment]
I quickly managed to get the matplotlib image via `BytesIO` into Emacs in a very basic way and with some drawbacks. See my branch emacs-plot-trial-1 here
https://github.com/brandm/live-py-plugin/tree/emacs-plot-trial-1

To try it out put this

```Python
import matplotlib.pyplot as plt
y = [0, 1 + 1, 9]
y *= 2
plt.plot(y)
```

into a Python buffer, `M-x live-py-mode RET`, go to the image buffer named `*live-py-image_[...]*`, `C-x 5 2` to create a new frame that can be resized individually with the image and go back to the Python buffer. Change the numbers to watch the effect on both the output buffer and the image buffer.

There seems to go quite some CPU somewhere (avoidable re-`import` or so?) and interaction can become annoying. It seems unavoidable for the call to code_tracer.py to become asynchronous. It would be very welcome if editing a simple Python code with for example `time.sleep(2)` would remain responsive before adding matplotlib support.

Then it would be good if generating the image could be decoupled from generating the output buffer so that the output buffer comes back to its original responsiveness and only the image takes a longer time. I hope that it is easy enough to process the common tasks like the above `1 + 1`, `*= 2` etc. and most of code_tracer.py only once.

Such a decoupling would probably also obsolete the following issue: stderr should probably not be misused for transporting the image because stderr could contain an important error message from code_tracer.py.

Also necessary at some point:

- As Don commented earlier, a choice between displaying only the output or only the image (or both for usage with extra frame): "switch back and forth between the visible buffers, so we have to pay attention to which buffer is visible when we update the display."

- Introduce `live-py-image-frame` or so for rebuilding `live-py-image-buffer` from scratch when in an own frame.

- emacs-plot-trial-1 is limited to the assumption that the Python code is always about a plot. To have code without a plot still working there should be some automatic detection or user interaction whether additionally an image buffer has to be updated and shown or not. It is not clear to me what and how.

- matplotlib should not be a requirement for non-plot usage.

**Comment by donkirkby** [comment]
That looks good, @brandm. I like your idea of returning the image bytes on stderr, and so far the errors have all been displayed as part of the display on stdout. Any error output from the driver or the traced code gets swallowed by the `FileSwallower`, we only display raised exceptions. Other options are temporary files or sockets, but stderr is much easier.

In the Eclipse plugin, I ran the code tracer in a background thread. That way, some slow code would be slow to update the live display, but the editor wouldn't slow down. Do you know of any way to run Emacs code in the background?

It's possible we could update the variable display in one thread and the plot display in another, but that seems more complicated than it's worth. Personally, I find the variable display useful for some code and the matplotlib display useful for other code. I don't know if I would use both on the same code.

**Comment by brandm** [comment]
Thanks for the feedback.

> Do you know of any way to run Emacs code in the background?

Emacs has [still no threading](https://lists.gnu.org/archive/html/emacs-devel/2016-10/msg00209.html) but there are options to run [asynchronous processes](http://www.gnu.org/software/emacs/manual/html_node/elisp/Asynchronous-Processes.html). For example for [Org Babel](http://orgmode.org/manual/Working-With-Source-Code.html) several attempts have been made to use it to run code blocks. [This example](https://github.com/astahlman/ob-async) has a nice demo.

**Comment by donkirkby** [comment]
An alternative to using stderr: use the [base64] module to encode the binary data from the image. Then we could report it like the turtle commands.

[base64]: https://docs.python.org/2/library/base64.html

**Comment by brandm** [comment]
base64 is not necessary at least for Emacs. Its buffer can handle 8 bit binary data inclusive null bytes like it does with `live-py-image-buffer` (try `C-c C-c` in this buffer). It should also be possible to add and strip MIME multipart separators still without base64. For now I would continue with stderr, maybe later MIME libraries can be used to pack and unpack all in stdout.

**Comment by donkirkby** [comment]
[MIME] sounds like a good idea. I like it!

[MIME]: https://docs.python.org/2/library/email.mime.html#email.mime.image.MIMEImage

**Comment by donkirkby** [comment]
Here's a promising [matplotlib tutorial] that could be adapted to live coding.

[matplotlib tutorial]: https://github.com/rougier/matplotlib-tutorial

**Comment by donkirkby** [comment]
Another source of inspiration: the [Python graph gallery].

[python graph gallery]: https://python-graph-gallery.com/

**Comment by donkirkby** [comment]
I'm marking this issue as `help wanted`, because I thought other people might have good suggestions on matplotlib tutorials to include. If you want to suggest a tutorial, you can just give a general description in a comment, or you could add a new issue with a description of what the student has to solve, a starting code snippet, and a target snippet. You can use other libraries like Seaborn, but I want to include at least some tutorials for basic matplotlib.

Here's a trivial example:

### Introduction ###
You can change the colour of a line by passing a colour abbreviation in the format string of the `plot()` method. Add a format string to these two `plot()` calls to make the heights red and the weights green.

### Start ###

    from matplotlib import pyplot as plt
    
    heights = [180, 180, 180, 173]
    weights = [72, 72, 64, 62]
    
    plt.plot(heights)
    plt.plot(weights)
    plt.show()

### Goal ###

    from matplotlib import pyplot as plt
    
    heights = [180, 180, 180, 173]
    weights = [72, 72, 64, 62]
    
    plt.plot(heights, 'r')
    plt.plot(weights, 'g')
    plt.show()

**Comment by donkirkby** [comment]
I moved the lessons feature into #176 and the request for help in designing lessons into #177.
This issue is now:

* [ ] Decide on a format for transmitting an image to the live canvas. Is there support for images on the `TurtleScreen` class? The other idea was to use MIME encoding.
* [ ] Monkey patch matplotlib to transmit an image from the `show()` command.
* [ ] Add a matplotlib example to the turtle example that appears when you launch the live canvas without any turtle commands.

**Comment by donkirkby** [comment]
This is now working in PyCharm. @ghost's example works if you add a call to `plt.show()`:

    import matplotlib.pyplot as plt
    import seaborn as sns
    sns.set(style="ticks")
    
    # Load the example dataset for Anscombe's quartet
    df = sns.load_dataset("anscombe")
    
    # Show the results of a linear regression within each dataset
    g = sns.lmplot(x="x", y="y", col="dataset", hue="dataset", data=df,
                   col_wrap=2, ci=None, palette="muted", size=4,
                   scatter_kws={"s": 50, "alpha": 1})#.getfig()
    
    plt.show()
